test_rmw_implementation/test/test_subscription.cpp

```diff
@@ -870,6 +871,143 @@ TEST_F(
+  });
+
+  size_t subscription_count = 0u;
+  SLEEP_AND_RETRY_UNTIL(rmw_intraprocess_discovery_delay, rmw_intraprocess_discovery_delay * 10) {
+    ret = rmw_publisher_count_matched_subscriptions(pub, &subscription_count);
+    if (RMW_RET_OK == ret && 1u == subscription_count) {  // Early return on failure.
+      break;
+    }
+  }
+
+  // Publish message with float (3.14159) from publisher to subscription
+  test_msgs__msg__BasicTypes original_message{};
+  ASSERT_TRUE(test_msgs__msg__BasicTypes__init(&original_message));
+  original_message.float32_value = 3.14159f;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    test_msgs__msg__BasicTypes__fini(&original_message);
+  });
+
+  rmw_publisher_allocation_t * null_allocation_p{nullptr};
+  rmw_subscription_allocation_t * null_allocation_s{nullptr};
+
+  ret = rmw_publish(pub, &original_message, null_allocation_p);
+  EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+
+  // expect to get message because no content filter set
+  {
+    rmw_subscriptions_t subscriptions;
+    void * subscriptions_storage[1];
+    subscriptions_storage[0] = sub->data;
+    subscriptions.subscribers = subscriptions_storage;
+    subscriptions.subscriber_count = 1;
+
+    rmw_wait_set_t * wait_set = rmw_create_wait_set(&context, 1);
+    ASSERT_NE(nullptr, wait_set) << rmw_get_error_string().str;
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+    {
+      EXPECT_EQ(
+        RMW_RET_OK, rmw_destroy_wait_set(wait_set)) << rmw_get_error_string().str;
+    });
+    rmw_time_t timeout = {1, 0};  // 1000ms
+    ret = rmw_wait(&subscriptions, nullptr, nullptr, nullptr, nullptr, wait_set, &timeout);
+    EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+    ASSERT_NE(nullptr, subscriptions.subscribers[0]);
+
+    test_msgs__msg__BasicTypes output_message{};
+    ASSERT_TRUE(test_msgs__msg__BasicTypes__init(&output_message));
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+    {
+      test_msgs__msg__BasicTypes__fini(&output_message);
+    });
+
+    ret = rmw_take(sub, &output_message, &taken, null_allocation_s);
+    // whatever the cft is supported or not
+    EXPECT_EQ(RMW_RET_OK, ret) << rmw_get_error_string().str;
+
+    EXPECT_TRUE(taken);
+    EXPECT_EQ(original_message, output_message);
+  }
+
+  // set content filter with 3.14
+  rmw_subscription_content_filter_options_t options =
+    rmw_get_zero_initialized_content_filter_options();
+  auto allocator = rcutils_get_default_allocator();
+  EXPECT_EQ(
+    RMW_RET_OK, rmw_subscription_content_filter_options_init(
+      "float32_value=3.14",
+      0,
+      nullptr,
+      &allocator,
+      &options));
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    EXPECT_EQ(
+      RMW_RET_OK,
+      rmw_subscription_content_filter_options_fini(&options, &allocator));
+  });
+
+  ret = rmw_subscription_set_content_filter(sub, &options);
+  if (ret != RMW_RET_UNSUPPORTED) {
+    ASSERT_EQ(RMW_RET_OK, ret);
```

Comment:
The test should sleep for a short amount (e.g. 500ms) to allow for the changed filter to propagate from the reader to the writer, otherwise there is a chance for the next published samples to be filtered by the old filter.

Reply:
I can add awaiting in the test, but I don't know if this API should be implemented as synchronous?
(NOTE: to wait in the test always causes a flaky error.)

Reply:
It seems there is still a limitation in this API, maybe `connextDDS` could have to add the reader filter side.
I think users don't want to get old filter data after calling `rmw_subscription_set_content_filter` for a subscription.

Reply:
I think that this is really important for user application.

this is said that a Writer side filter may be using outdated expression parameters.
if this is negative case, that should be okay because filter signature does not match by Reader side, the Reader will ignore the filter results and instead apply its local filter?
but if it is positive case, what would happen? since the message is already dropped by writer side with outdated filter expression?

Reply:
> if this is negative case, that should be okay because filter signature does not match by Reader side, the Reader will ignore the filter results and instead apply its local filter?

Yes, I think so.

> if it is positive case, what would happen? since the message is already dropped by writer side with outdated filter expression?

IMO, users can accept that the message is dropped by writer side in this case.

Reply:
> IMO, users can accept that the message is dropped by writer side in this case.

I just wondering that if i am developing application with CFT, i want to know when exactly the filtering expression and parameters are in effect even on DataWriter. instead, we have no idea outdated filtering still in effect... 

@asorbini @MiguelCompany any thoughts?